gpflow/posterior.py

```diff
@@ -0,0 +1,194 @@
+        Kuu = covariances.Kuu(self.iv, self.kernel, jitter=default_jitter())  # [(R), M, M]
+        q_mu = self.q_dist.q_mu
+
+        if Kuu.shape.ndims == 4:
+            ML = tf.cast(tf.sqrt(tf.cast(tf.reduce_prod(tf.shape(Kuu)), default_float())), tf.int32)
+            Kuu = tf.reshape(Kuu, [ML, ML])
+        if Kuu.shape.ndims == 3:
+            q_mu = tf.linalg.adjoint(self.q_dist.q_mu)[..., None]  # [..., R, M, 1]
+        L = tf.linalg.cholesky(Kuu)
+
+        if not self.whiten:
+            # alpha = Kuu⁻¹ q_mu
+            alpha = tf.linalg.cholesky_solve(L, q_mu)
+        else:
+            # alpha = L⁻T q_mu
+            alpha = tf.linalg.triangular_solve(L, q_mu, adjoint=True)
+        # predictive mean = Kfu alpha
+        # predictive variance = Kff - Kfu Qinv Kuf
+        # S = q_sqrt q_sqrtT
+        if not self.whiten:
+            # Qinv = Kuu⁻¹ - Kuu⁻¹ S Kuu⁻¹
+            #      = Kuu⁻¹ - L⁻T L⁻¹ S L⁻T L⁻¹
+            #      = L⁻T (I - L⁻¹ S L⁻T) L⁻¹
+            #      = L⁻T B L⁻¹
+            if isinstance(self.q_dist, DiagNormal):
+                q_sqrt = tf.linalg.diag(tf.linalg.adjoint(self.q_dist.q_sqrt))
+            else:
+                q_sqrt = self.q_dist.q_sqrt
+            Linv_qsqrt = tf.linalg.triangular_solve(L, q_sqrt)
+            Linv_cov_u_LinvT = tf.matmul(Linv_qsqrt, Linv_qsqrt, transpose_b=True)
+        else:
+            if isinstance(self.q_dist, DiagNormal):
+                Linv_cov_u_LinvT = tf.linalg.diag(tf.linalg.adjoint(self.q_dist.q_sqrt ** 2))
+            else:
+                q_sqrt = self.q_dist.q_sqrt
+                Linv_cov_u_LinvT = tf.matmul(q_sqrt, q_sqrt, transpose_b=True)
+            # Qinv = Kuu⁻¹ - L⁻T S L⁻¹
+            # Linv = (L⁻¹ I) = solve(L, I)
+            # Kinv = Linv.T @ Linv
+        I = tf.eye(tf.shape(Linv_cov_u_LinvT)[-1], dtype=Linv_cov_u_LinvT.dtype)
+        B = I - Linv_cov_u_LinvT
+        LinvT_B = tf.linalg.triangular_solve(L, B, adjoint=True)
+        B_Linv = tf.linalg.adjoint(LinvT_B)
+        Qinv = tf.linalg.triangular_solve(L, B_Linv, adjoint=True)
+
+        return alpha, Qinv
+
+    def update_cache(self):
+        self.alpha, self.Qinv = self._precompute()
+
+    def freeze(self):
+        alpha, Qinv = self._precompute()
+        self.alpha = Parameter(alpha, trainable=False)
+        self.Qinv = Parameter(Qinv, trainable=False)
+
+    def update_cache_with_variables(self):
+        alpha, Qinv = self._precompute()
+        if isinstance(self.alpha, Parameter) and isinstance(self.Qinv, Parameter):
+            self.alpha.assign(alpha)
+            self.Qinv.assign(Qinv)
+        else:
+            self.alpha = Parameter(alpha, trainable=False)
+            self.Qinv = Parameter(Qinv, trainable=False)
+
+    def predict_f(
+        self, Xnew, full_cov: bool = False, full_output_cov: bool = False
+    ) -> MeanAndVariance:
+        # Qinv: [L, M, M]
+        # alpha: [M, L]
+
+        Kuf, Knn, fully_correlated = _get_kernels(
```

Comment:
Nit: is `_get_kernels` the most precise name? It looks to me like we are computing covariance matrices?